### PR TITLE
fix: use purl as bomref

### PR DESCRIPTION
**Pull request by jeremylong**
For the metadata component it would be great if we had a purl. As such, if we just make the bomref a purl we can use the value for both the bomref and the purl.

### CODE REVIEW

**Comment by macblazer** [comment]
The metadata component is the thing being analyzed.  `pkg:cocoapods/` is the purl prefix to use for packages that are available to be imported with CocoaPods.  If the thing being analyzed is, itself, a Pod then this is cool.

However, normally the thing being analyzed is an application being built that uses CocoaPods as its dependency manager.  So it should maybe have something like an `app:...` prefix, although I am not aware if there are any standards around this.

**Comment by jeremylong** [comment]
I'll look into adding a [purl type](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst) as an option that defaults to `generic`.

**Comment by jeremylong** [comment]
After looking at the code some more - I think we can infer the package type from the specified component type. The purl type should be ' generic ' if the component type is anything but `library`. If the component type is `library` then the purl type should be `cocoapods`. If you want, we could add another option to override this behavior.

File: lib/cyclonedx/cocoapods/component.rb [hunk 1]
```diff
@@ -53,13 +53,24 @@ def initialize(name:, version:, type:, group: nil, build_system: nil, vcs: nil)
...
-        return if group.nil?
+      private
 
-        @bomref = "#{group}/#{@bomref}"
+      def build_purl(type, name, group, version)
+        package_type = type == 'library' ? 'cocoapods' : 'generic'
+        if group.nil?
+          purl_name, subpath = parse_name(name)
+        else
+          # this seems wrong as cocoapods does not use groups?
```
Comment (macblazer) [inline lib/cyclonedx/cocoapods/component.rb:69]: This is the part I am not comfortable with for a `library` style `purl` here.  If we're analyzing an application, then `pkg:generic/<group>/<name>@<version>#<subpath>` is correct I think.  The caller enters the group, name, and version as part of the CLI to invoke this, and then we just say it's a `pkg:generic` purl that may or may not be in anybody else's databases.

However, if the thing we are analyzing is itself a `library` that is available through `CocoaPods` and if the `bom-ref` is a `purl`, I would expect it to be the correct purl of the pod being analyzed and **not** up to the CLI parameters being passed.  The information to build the `purl` would be in a `*.podspec` file where the name of the file is the name of the pod.  We don't currently have code here to read a `podspec` file, but it is part of the ruby library that we are importing from the CocoaPods project.

**Comment by jeremylong** [comment]
@macblazer I think this is ready for your re-review now. I believe I have addressed your concerns about using the package type of cocopods with [9fda73d](https://github.com/CycloneDX/cyclonedx-cocoapods/pull/84/commits/9fda73de5e976115c382baeef9919b0f50777c71) and I have added a test with [ce9bcaf](https://github.com/CycloneDX/cyclonedx-cocoapods/pull/84/commits/ce9bcaf55afccf59a06f9e74ed29c0a59c9c12fb). Yes, this is abusing the 'type' slightly, but I felt this was better than upping the parameter count to 7.

File: lib/cyclonedx/cocoapods/component.rb [hunk 2]
```diff
@@ -74,13 +85,39 @@ def validate_attributes(name, version, type, group)
...
+
       def missing(str)
         str.nil? || str.to_s.strip.empty?
       end
 
       def exists_and_blank(str)
         !str.nil? && str.to_s.strip.empty?
       end
+
+      def create_purl(name)
```
Comment (macblazer) [inline lib/cyclonedx/cocoapods/component.rb:109]: This `create_purl` function and the `parse_name` function above seem to be identical.  Is this function called anywhere?

**Review by macblazer — CHANGES_REQUESTED** [review]
Looks good other than the one duplicate function that seems unused.  I like this ability to read the podspec file and generate the purl of the thing being analyzed if it is a Pod.

File: lib/cyclonedx/cocoapods/component.rb (hunk 2 above)
Comment (jeremylong) [reply lib/cyclonedx/cocoapods/component.rb]: The duplicate, dead code has been removed.

**Review by macblazer — APPROVED** [review]

**Pull request merged**